_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,883 @@
+      IO{generator match
+            case Hotp =>
+                s"otpauth://hotp/$issuer:$account?secret=$secretKey&issuer=$issuer&algorithm=SHA256&counter=0"
+            case Totp =>
+                    s"otpauth://totp/$issuer:$account?secret=$secretKey&issuer=$issuer&algorithm=SHA256"
+      }
+
+    def createQRCode(barCodeData: IO[String], filePath: String = "barCode.png", height: Int = 400, width: Int = 400): IO[Either[String, Unit]] =
+        barCodeData.map{data =>
+                Try{
+                    val matrix: BitMatrix = new MultiFormatWriter().encode(data, BarcodeFormat.QR_CODE, width, height)
+                    val outVal: FileOutputStream = new FileOutputStream(filePath)
+                    MatrixToImageWriter.writeToStream(matrix,"png",outVal)
+                }.toEither.left.map(_.getMessage)
+            }
+}
+```
+
+**Creating the OtpService.**
+
+Let's create a scala file and save it as `OtpService.scala`. This service will contain the implementation for `Hotp` and `Totp` that we will use for this application.
+
+```scala
+package com.rockthejvm
+
+import scala.util.*
+import com.rockthejvm.Generator
+
+class OtpService(generator: Generator, user: User){
+    private val secret: Array[Byte] = SecretGenerator.generate()
+}
+```
+
+We create it as a class in order to pass a `Generator` and `User` as constructor arguments during instantiation. A new `secret` value is generated by otp-java every time a user tries to sign in and saved as an `Array[Byte]` in a private val.
+
+```scala
+import com.bastiaanjansen.otp.*
+import java.time.Duration
+
+class OtpService(generator: Generator, user: User){
+    ...
+    private object GenInstance {
+        def hotpGen(secretValue: Array[Byte]) = new HOTPGenerator.Builder(secretValue)
+            .withPasswordLength(6)
+            .withAlgorithm(HMACAlgorithm.SHA256)
+            .build()
+        def totpGen(secretValue: Array[Byte]) = new TOTPGenerator.Builder(secretValue)
+                .withHOTPGenerator(builder => {
+                        builder.withPasswordLength(6)
+                        builder.withAlgorithm(HMACAlgorithm.SHA256)
+                })
+                .withPeriod(Duration.ofSeconds(30))
+                .build()
+    }
+}
+```
+
+The `GenInstance` object contains two functions `hotpGen()` and `totpGen()` used to provide instances of `HOTPGenerator` and `TOTPGenerator` respectively. They both take a `secret` as a parameter. Note that Google Authenticator displays a token length of 6, we supply the same value to our function implementations. We also use the same algorithm type `SHA256` as our `URI`. This object is created as `private` so that it's only used within this class.
+
+```scala
+import cats.effect.IO
+import com.rockthejvm.Generator.*
+
+class OtpService(generator: Generator, user: User){
+    ...
+    def getToken(generator: Generator = generator, counter: Long = user.counterValue): IO[Either[String,String]] =
```

Comment:
these arguments aren't necessary - generator is already in the service, the counter can be extracted from the user

Reply:
Updated